app/controllers/authenticate_controller.rb

```diff
@@ -1,30 +1,89 @@
 class AuthenticateController < ApplicationController
   include TokenGenerator
 
-  # Request path indicates the username for +authenticate+.
-  before_filter :credentials_lookup
+  class MemoizedRole
+    def self.[](role_id)
+      @user_roles ||= Hash.new { |h, id| h[id] = Role[id] }
+      @user_roles[role_id]
+    end
+
+    def self.roleid_from_username(account, username)
+      Role.roleid_from_username(account, username)
+    end
+  end
 
   def authenticate
-    api_key = request.body.read
-    if @credentials.valid_api_key? api_key
-      render json: authentication_token
+
+    authenticator  = params[:authenticator]
+    service_id     = params[:service_id]
+    account        = params[:account]
+    username       = params[:id]
+    password       = request.body.read
+    role_id        = MemoizedRole.roleid_from_username(account, username)
+
+    validate_role_exists!(role_id)
+    
+    case authenticator
+    when 'authn-ldap'
```

Comment:
all custom authn are going to have to  `validate_security!`, so it would be nice if this was written so that everyone adding a custom authn didn't have to remember to do that